### ISSUE TITLE
Bit-for-bit deterministic / reproducible builds

### ISSUE THREAD

**Issue by infinity0** [rust-lang/rust#34902]
It would be good if rustc could generate bit-for-bit reproducible results, even in the presence of minor system environment differences. [Currently](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/rustc.html) we have quite a large diff: e.g. see [txt diff for 1.9.0](https://tests.reproducible-builds.org/debian/dbdtxt/unstable/amd64/rustc_1.9.0+dfsg1-1.diffoscope.txt) or perhaps [txt diff for 1.10.0](https://tests.reproducible-builds.org/debian/dbdtxt/unstable/amd64/rustc_1.10.0+dfsg1-1.diffoscope.txt) a few days after I'm posting this. (You might want to "save link as" instead of displaying it directly in the browser.)

Much of the diff output is due to build-id differences, which can be ignored since they are caused by other deeper issues and will go away once these deeper issues are fixed. One example of a deeper issue is this:

```
│   │   │   ├── ./usr/lib/i386-linux-gnu/librustc_data_structures-e8edd0fd.so
│   │   │   │   ├── readelf --wide --file-header {}
│   │   │   │   │ @@ -6,15 +6,15 @@
│   │   │   │   │    OS/ABI:                            UNIX - System V
│   │   │   │   │    ABI Version:                       0
│   │   │   │   │    Type:                              DYN (Shared object file)
│   │   │   │   │    Machine:                           Intel 80386
│   │   │   │   │    Version:                           0x1
│   │   │   │   │    Entry point address:               0x4620
│   │   │   │   │    Start of program headers:          52 (bytes into file)
│   │   │   │   │ -  Start of section headers:          338784 (bytes into file)
│   │   │   │   │ +  Start of section headers:          338780 (bytes into file)
│   │   │   │   │    Flags:                             0x0
│   │   │   │   │    Size of this header:               52 (bytes)
│   │   │   │   │    Size of program headers:           32 (bytes)
│   │   │   │   │    Number of program headers:         8
│   │   │   │   │    Size of section headers:           40 (bytes)
│   │   │   │   │    Number of section headers:         30
│   │   │   │   │    Section header string table index: 29
│   │   │   │   ├── readelf --wide --program-header {}
```

[Here](https://tests.reproducible-builds.org/debian/index_variations.html) are the system variations that might be causing these issues. I myself am not that familiar with ELF, but perhaps someone else here would know why the section header is 4 bytes later in the first vs second builds.

**Comment by steveklabnik** [comment]
In general, being reproducible is something we're interested in; we try to tackle it bug by bug.

**Comment by eefriedman** [comment]
This probably has some obvious cause:
```
│   │   │   │ -<h4 id='method.unzip' class='method'><code>fn <a href='../../core/iter/trait.Iterator.html#method.unzip' class='fnname'>unzip</a>&lt;A, B, FromA, FromB&gt;(self) -&gt; (FromA, FromB) <span class='where'>where FromB: <a class='trait' href='../../core/default/trait.Default.html' title='core::default::Default'>Default</a> + <a class='trait' href='../../core/iter/trait.Extend.html' title='core::iter::Extend'>Extend</a>&lt;B&gt;, FromA: <a class='trait' href='../../core/default/trait.Default.html' title='core::default::Default'>Default</a> + <a class='trait' href='../../core/iter/trait.Extend.html' title='core::iter::Extend'>Extend</a>&lt;A&gt;, Self: <a class='trait' href='../../core/iter/trait.Iterator.html' title='core::iter::Iterator'>Iterator</a>&lt;Item=(A, B)&gt;</span></code></h4>
│   │   │   │ +<h4 id='method.unzip' class='method'><code>fn <a href='../../core/iter/trait.Iterator.html#method.unzip' class='fnname'>unzip</a>&lt;A, B, FromA, FromB&gt;(self) -&gt; (FromA, FromB) <span class='where'>where FromB: <a class='trait' href='../../core/default/trait.Default.html' title='core::default::Default'>Default</a> + <a class='trait' href='../../core/iter/trait.Extend.html' title='core::iter::Extend'>Extend</a>&lt;B&gt;, Self: <a class='trait' href='../../core/iter/trait.Iterator.html' title='core::iter::Iterator'>Iterator</a>&lt;Item=(A, B)&gt;, FromA: <a class='trait' href='../../core/default/trait.Default.html' title='core::default::Default'>Default</a> + <a class='trait' href='../../core/iter/trait.Extend.html' title='core::iter::Extend'>Extend</a>&lt;A&gt;</span></code></h4>
```

If someone wants to look at the code generation differences, it's probably best to start with libcore.  The diff you linked isn't that useful for that because it doesn't recognize .rlib files as archives.

**Comment by infinity0** [comment]
Thanks for the heads up, I [just added AR support in diffoscope](https://anonscm.debian.org/cgit/reproducible/diffoscope.git/commit/?id=9415dd7d5ea7be1d65e506f06b31ad205b4bcbbb) and hopefully the diff output will see that in the next few weeks, when the website updates.

**Comment by infinity0** [comment]
Here's a [diff](https://people.debian.org/~infinity0/res/libcore.diff.html) of rust.metadata.bin from libcore.rlib, from 1.10.0. (2.0 MB, ~1700 lines, might make your browser slow.) Mostly ordering differences. @eddyb has suggested HashMap ordering as a culprit, and that would be replaced with FnvHashMap instead.

**Comment by eddyb** [comment]
If it's a `HashMap`, I don't know *which* one. cc @rust-lang/compiler

**Comment by infinity0** [comment]
I tried replacing HashMap with FnvHashMap in `src/librustc_metadata/loader.rs` but it didn't seem to help. Here is the Makefile i'm using to automate my rebuilds:

~~~~
TRIPLET = x86_64-unknown-linux-gnu

all: libcore.diff

libcore.diff: rust.metadata.bin.1 rust.metadata.bin.2
	diffoscope --html-dir "$@" rust.metadata.bin.1 rust.metadata.bin.2 \
		--max-diff-block-lines 5000 \
		--max-diff-input-lines 10000000 \
		--max-report-size 204800000; true # diffoscope returns 1 for "there were diffs"

rust.metadata.bin.1 rust.metadata.bin.2: Makefile
	rm -f $(TRIPLET)/stage2/lib/rustlib/$(TRIPLET)/lib/libcore-*.rlib
	rm -f $(TRIPLET)/stage2/lib/rustlib/$(TRIPLET)/lib/stamp.core
	$(MAKE) $(TRIPLET)/stage2/lib/rustlib/$(TRIPLET)/lib/stamp.core
	ar x $(TRIPLET)/stage2/lib/rustlib/$(TRIPLET)/lib/libcore-*.rlib rust.metadata.bin
	mv rust.metadata.bin "$@"

.PHONY: all rust.metadata.bin.1 rust.metadata.bin.2
~~~~

perhaps i'm Doing It Wrong.

**Comment by infinity0** [comment]
As background, HashMap in rust is non-deterministic to protect against certain types of DoS attack. You can switch it to the deterministic FnvHashMap if you're sure your code will always be called in a safe manner. This ought to be true for rustc itself. (I notice some online "try-it-yourself" rust web services let me run "ls /" and other shell commands, so I could also exploit this HashMap ordering issue, but I also assume that they're clever enough to set a ulimit and/or containerise the thing.)

**Comment by jonas-schievink** [comment]
I wrote a small [script](https://gist.github.com/jonas-schievink/18cf82523262b8ca62f94d54bef6f905) to diff metadata (I couldn't really get the makefile to work).

It looks like even more is changing between two compiles using the current master (or nightly): The crate hash and/or disambiguator, which are stored right after the target triple. @infinity0 can you confirm?

**Comment by jonas-schievink** [comment]
Okay, apparently replacing `Hash{Map,Set}` in resolve with `FnvHash{Map,Set}` got rid of the first (smaller) blocks of the diff. The big block at the bottom is still there. Will continue to randomly `sed` the compiler and report back ;)

**Comment by jonas-schievink** [comment]
The remaining issue is that (some?) predicates are encoded in non-deterministic order. Maybe we need to do [this](https://github.com/rust-lang/rust/blob/3c5a0fa45b5e2786b6e64e27f48cd129e7aefdbd/src/librustc_typeck/collect.rs#L2019) for all bounds? Not sure what would be the correct way to do that (or why these are nondeterministic in the first place).

**Comment by michaelwoerister** [comment]
Just for reference, there's this: https://github.com/rust-lang/rust/pull/34805

**Comment by jonas-schievink** [comment]
@michaelwoerister Thanks! Too tired to think about it currently. Will have a look tomorrow :)

**Comment by sanxiyn** [comment]
#24473 is rustdoc subset of this.

**Comment by nikomatsakis** [comment]
Potentially relevant, from IRC:

```
 <eddyb> nmatsakis, infinity0: fwiw, I just found another hash table. not random, but may cause issues around incremental recompilation. see rustc_metadata::encoder::encode_reachable (the reachable NodeSet)
```

**Comment by eddyb** [comment]
Fix suggestion is changing that to be a `BTreeSet<DefIndex>`. Other things could also use btrees.

**Comment by infinity0** [comment]
In case this helps to motivate anyone, we got a successful reproduction on i386 on Debian testing!

https://tests.reproducible-builds.org/debian/rb-pkg/testing/i386/rustc.html

This may or may not be "an accident", we'll have to see what future tests show. Also on this arch/platform we are fixing the build path (just to see how it does). For other arch/platforms we vary the build path, and haven't seen rustc reproduce there yet.

**Comment by michaelwoerister** [comment]
At least when building with debuginfo, the build path will show up in the resulting binaries as the `DW_AT_comp_dir` attribute of the `DW_TAG_compile_unit`.

**Comment by jmesmon** [comment]
@michaelwoerister 

In gcc (and clang), the `-fdebug-prefix-map` argument allows one to avoid this. Is a similar option for rustc available?

**Comment by michaelwoerister** [comment]
@jmesmon No, we don't have an option like that. If you open an issue, I'd be happy to discuss it with the @rust-lang/tools team.

**Comment by infinity0** [comment]
Assuming that #41508 works properly, there are only two issues left, judging from [this diff](https://tests.reproducible-builds.org/debian/dbdtxt/testing/amd64/rustc_1.14.0+dfsg1-3.diffoscope.txt) obtained when not varying the build path across builds.

- #24473 for rustdoc, with several instances of 
```diff
├── ./usr/share/doc/rust-doc/html/implementors/core/cmp/trait.Ord.js
├── js-beautify {}
@@ -1,10 +1,9 @@
 (function() {
     var implementors = {};
-    implementors["alloc"] = ["impl&lt;T:&nbsp;? [.. snip ..]
     implementors["rustc_unicode"] = [];
     implementors["collections"] = ["impl&lt;T&gt; [.. snip ..]
     implementors["core"] = [];
 
     if (window.register_implementors) {
         window.register_implementors(implementors);
     } else {
```
- one non-build-path debuginfo variation:
```diff
     <2ba6>   DW_AT_GNU_dwo_name: (indirect string, offset: 0x33a0): x86_64-unknown-linux-gnu/rustllvm/PassWrapper.dwo
     <2baa>   DW_AT_comp_dir    : .
     <2bac>   DW_AT_GNU_pubnames: 1
     <2bac>   DW_AT_GNU_addr_base: 0x4d90
-    <2bb0>   DW_AT_GNU_dwo_id  : 0xeed3d315f58a4192
+    <2bb0>   DW_AT_GNU_dwo_id  : 0xe31c7153633c9772
     <2bb8>   DW_AT_GNU_ranges_base: 0x57f0
   Compilation Unit @ offset 0x2bbc:
```

**Comment by infinity0** [comment]
The doc issue seems to be gone, the dwo_id issue [remains](https://tests.reproducible-builds.org/debian/dbdtxt/buster/amd64/rustc_1.16.0+dfsg1-1.diffoscope.txt).

```diff
@@ -4968,40 +4968,40 @@
     <2ef2>   DW_AT_ranges      : 0x51c0
     <2ef6>   DW_AT_low_pc      : 0x0
     <2efe>   DW_AT_stmt_list   : 0x1110
     <2f02>   DW_AT_GNU_dwo_name: (indirect string, offset: 0x3221): x86_64-unknown-linux-gnu/rustllvm/RustWrapper.dwo
     <2f06>   DW_AT_comp_dir    : .
     <2f08>   DW_AT_GNU_pubnames: 1
     <2f08>   DW_AT_GNU_addr_base: 0x0
-    <2f0c>   DW_AT_GNU_dwo_id  : 0x82dcd4a4a86511a
+    <2f0c>   DW_AT_GNU_dwo_id  : 0x94cc2c8c89838e78
     <2f14>   DW_AT_GNU_ranges_base: 0x1290
   Compilation Unit @ offset 0x2f18:
    Length:        0x2e (32-bit)
    Version:       4
    Abbrev Offset: 0x206
    Pointer Size:  8
  <0><2f23>: Abbrev Number: 1 (DW_TAG_compile_unit)
     <2f24>   DW_AT_ranges      : 0x8250
     <2f28>   DW_AT_low_pc      : 0x0
     <2f30>   DW_AT_stmt_list   : 0x4366
     <2f34>   DW_AT_GNU_dwo_name: (indirect string, offset: 0x3253): x86_64-unknown-linux-gnu/rustllvm/PassWrapper.dwo
     <2f38>   DW_AT_comp_dir    : .
     <2f3a>   DW_AT_GNU_pubnames: 1
     <2f3a>   DW_AT_GNU_addr_base: 0x4f08
-    <2f3e>   DW_AT_GNU_dwo_id  : 0x6564472d68267dc6
+    <2f3e>   DW_AT_GNU_dwo_id  : 0xa4b7031387acdfba
     <2f46>   DW_AT_GNU_ranges_base: 0x58a0
   Compilation Unit @ offset 0x2f4a:
    Length:        0x2e (32-bit)
    Version:       4
    Abbrev Offset: 0x223
    Pointer Size:  8
  <0><2f55>: Abbrev Number: 1 (DW_TAG_compile_unit)
     <2f56>   DW_AT_ranges      : 0xaf70
     <2f5a>   DW_AT_low_pc      : 0x0
     <2f62>   DW_AT_stmt_list   : 0x6382
     <2f66>   DW_AT_GNU_dwo_name: (indirect string, offset: 0x3285): x86_64-unknown-linux-gnu/rustllvm/ArchiveWrapper.dwo
     <2f6a>   DW_AT_comp_dir    : .
     <2f6c>   DW_AT_GNU_pubnames: 1
     <2f6c>   DW_AT_GNU_addr_base: 0x7528
-    <2f70>   DW_AT_GNU_dwo_id  : 0xc1d2b489290c9d87
+    <2f70>   DW_AT_GNU_dwo_id  : 0xcd75544460bc0e9c
     <2f78>   DW_AT_GNU_ranges_base: 0x8480
```

**Comment by michaelwoerister** [comment]
This `DW_AT_GNU_dwo_id` seems to be introduced by the C++ compiler when compiling `rustllvm`. @alexcrichton, can we disable split debuginfo here somehow?

**Comment by michaelwoerister** [comment]
cc @brson

**Comment by infinity0** [comment]
Also FYI on Debian (where these tests are done) rustllvm is dynamically linked against the LLVM shared library, in case this affects the situation.

**Comment by alexcrichton** [comment]
@michaelwoerister sure yeah we can just pass whatever flags are necessary to the C compiler

**Comment by infinity0** [comment]
It looks like `llvm-config -cxxflags` on Debian at least outputs `-gsplit-dwarf` so I'm applying this patch for the time being:

```diff
--- a/src/librustc_llvm/build.rs
+++ b/src/librustc_llvm/build.rs
@@ -137,6 +137,11 @@
     let cxxflags = output(&mut cmd);
     let mut cfg = gcc::Config::new();
     for flag in cxxflags.split_whitespace() {
+        // Split-dwarf gives unreproducible DW_AT_GNU_dwo_id so don't do it
+        if flag == "-gsplit-dwarf" {
+            continue;
+        }
+
         // Ignore flags like `-m64` when we're doing a cross build
         if is_crossed && flag.starts_with("-m") {
             continue;
```

Not sure if rustc wants to include it, since technically it is an issue with either `llvm-config` or the C++ compiler.

**Comment by infinity0** [comment]
I was hoping to be able to declare rustc the first reproducible compiler, but unfortunately it seems even with the above patch there was a regression between 1.18 and 1.19, see [libcore.diff](https://people.debian.org/~infinity0/rust/libcore.diff/) which was produced using the same build-path. Probably another case of non-deterministic hash tables being used; I will see if I can write a test case to avoid this happening again in the future.

For other diffs as well as the actual rustc binaries see [the parent directory](https://people.debian.org/~infinity0/rust/).

**Comment by infinity0** [comment]
Oh, I just had a thought, I added `codegen-units = 0` recently to our `config.toml` which enables parallel codegen. This might be interfering with reproducibility; I will try disabling that and re-running the builds.

**Comment by infinity0** [comment]
OK! After disabling parallel codegen we get much much further, but there is still a regression, [the build of liballoc_jemalloc](https://people.debian.org/~infinity0/rust/20170801/libstd-rust-dev.diff/) is no longer deterministic.

tests.r-b.org has also caught up with us and is [agreeing with this result](https://tests.reproducible-builds.org/debian/rb-pkg/buster/i386/diffoscope-results/rustc.html), though a [bug in diffoscope](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=870049) that got fixed recently causes it not to display any output at the time of writing (it should be fixed automatically soon).

This is the [jemalloc diff](https://people.debian.org/~infinity0/rust/20170801/rustc_1.17.0_1.19.0_liballoc_jemalloc.diff) between 1.17 and 1.19. Is it worth me trying to chase this down or were you guys planning to switch the default to the system allocator soon? I seem to remember an ongoing discussion about that.

**Comment by alexcrichton** [comment]
Thanks so much for the ongoing work here @infinity0! 

@michaelwoerister do you know if `codegen-units > 1` would cause the build to be non-reproducible? I figured that w/ incremental work that wouldn't be the case!

@infinity0 we're unfortunately still aways away from jettisoning jemalloc. We need the `#[global_allocator]` attribute to stabilize first before doing so, which would still be at least 12 weeks away (ish) if we stabilized it today. We don't currently have a timeline for the stabilization of that yet, but it should be relatively close in the sense that the design is much closer to "stabilizable" than the previous.

**Comment by alexcrichton** [comment]
> Cargo seems to build things in random order -- that could affect the linker flags, right?

Hm perhaps! I think we did a change to Cargo awhile ago to make any one rustc invocation determinstic, e.g. https://github.com/rust-lang/cargo/pull/3937

**Comment by infinity0** [comment]
With 1.21.0+dfsg1-3 Debian, rustc the compiler is built deterministically (and build-path-independently) but libstd has [some things remaining](https://tests.reproducible-builds.org/debian/dbdtxt/unstable/amd64/rustc_1.21.0+dfsg1-3.diffoscope.txt):

```patch
./usr/lib/x86_64-linux-gnu/librustc-2d0e4ac759a96c13.so
objdump --line-numbers --disassemble --demangle --section=.text {}
@@ -39978,25 +39978,25 @@
    26491:    48 89 e5                mov    %rsp,%rbp
 /usr/src/rustc-1.21.0/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/src/float/sub.rs:11
    26494:    f2 0f 5c c1             subsd  %xmm1,%xmm0
 /usr/src/rustc-1.21.0/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/src/macros.rs:255
    26498:    5d                      pop    %rbp
    26499:    c3                      retq   
    2649a:    66 90                   xchg   %ax,%ax
-/build/1st/rustc-1.21.0+dfsg1/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:39
+/build/rustc-1.21.0+dfsg1/2nd/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:39
    2649c:    66 0f 6e c7             movd   %edi,%xmm0
-/build/1st/rustc-1.21.0+dfsg1/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:40
+/build/rustc-1.21.0+dfsg1/2nd/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:40
    264a0:    48 c1 ef 20             shr    $0x20,%rdi
-/build/1st/rustc-1.21.0+dfsg1/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:41
+/build/rustc-1.21.0+dfsg1/2nd/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:41
    264a4:    48 0b 3d 25 19 00 00    or     0x1925(%rip),%rdi        
-/build/1st/rustc-1.21.0+dfsg1/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:42
+/build/rustc-1.21.0+dfsg1/2nd/src/rustc/compiler_builtins_shim/../../libcompiler_builtins/compiler-rt/lib/builtins/x86_64/floatundidf.S:42
    264ab:    66 0f 56 05 fd 18 00    orpd   0x18fd(%rip),%xmm0        
    264b2:    00 
```

What's strange is that some of the paths are already remapped to `/usr/src` as I intended with the `-Zremap-path-prefix` flags I'm passing. However, I'm only passing `-Zremap-path-prefix-from=$PWD`, but some other experiments with @kpcyrd on other rust crates suggests that we need to also pass `-Zremap-path-prefix-from=$CARGO_HOME` in the general case. I'll try that next time and report back, hopefully everything will be reproducible then.

**Comment by infinity0** [comment]
I should mention that I had to disable jemalloc in Debian for other reasons, so builds with that enabled might still hit the nondeterminism I [mentioned earlier](https://github.com/rust-lang/rust/issues/34902#issuecomment-319463586).

**Comment by arielb1** [comment]
@infinity0 

Nice to hear we almost got rid of nondeterminism! The last bunch of path dependence looks like it comes from [compiler-rt] asm files that are a part of [compiler-builtins] - I think the build-script that compiles them all is [here].

[compiler-builtins]: http://github.com/rust-lang-nursery/compiler-builtins 
[compiler-rt]: https://github.com/rust-lang/compiler-rt
[here]: https://github.com/rust-lang-nursery/compiler-builtins/blob/f5532b22b5d741f3ea207b5b07e3e1ca63476f9b/build.rs#L4435

**Comment by daym** [comment]
Rust 1.22.1 has a reproducibility problem in liblibc's rust.metadata.bin .The symbols in there seems to be shuffled (example: `getutxent`).

**Comment by michaelwoerister** [comment]
@daym I can't find anything in the metadata encoder that immediately seems at error here. Would be interesting to look into this some more.

**Comment by daym** [comment]
@michaelwoerister Thanks for checking! With newer rusts, the situation is much better. We work around the remaining reproducibility problems of rust 1.25, rust 1.26, rust 1.27 by using llvm 3.9.1 instead of llvm 6.0.1. Even newer rusts are fine.

So in principle the fix could be bisected.

But we use rust 1.22.1 as part of a bootstrapping effort, bootstrapping the rust compiler from C++ via mrustc.

mrustc so far only supports compiling the rust 1.19.0 compiler, so that's how we ended up (eventually) caring about rust 1.22.1. That said, maybe just wait until mrustc supports compiling the rust 1.23.0 compiler directly. We'll see which comes first.

**Comment by elichai** [comment]
What's the status of this? Is it already possible to make a reproducible/deterministic build or not?
(I'm interested in both binaries and staticlibs)

**Comment by tarcieri** [comment]
My understanding is builds are now (sometimes?) reproducible and that people are using [reprotest](https://pypi.org/project/reprotest/) successfully in CI.

To me it seems like we're not too far away from robust reproducible builds, but the main problem right now is there is no easy-to-use tooling for producing and verifying such builds.

I opened an issue about that on the Secure Code WG issue tracker:

https://github.com/rust-secure-code/wg/issues/28

**Comment by zmanian** [comment]
Isn't this a significant milestone on reproducibility.

https://www.reddit.com/r/rust/comments/afscgo/ripgrep_0100_is_reproducible_in_debian/

**Comment by dwijnand** [comment]
> My understanding is builds are now (sometimes?) reproducible and that people are using [reprotest](https://pypi.org/project/reprotest/) successfully in CI.

@tarcieri can you share some links of CI setups for that?

**Comment by tarcieri** [comment]
Well, if you're ok with cargo culting a config from a project called `sniffglue`, here is an example:

https://github.com/kpcyrd/sniffglue/blob/master/ci/reprotest.sh

**Comment by sanxiyn** [comment]
Just to prevent some confusion: these two issues are separate: 1) reproducible builds of softwares written in Rust, and 2) reproducible builds of Rust compiler. Since Rust compiler is written in Rust, 2 requires 1.

Since this is an issue on Rust compiler repository, this issue is mostly about 2. 2 is very close but not done. So this issue is open.

1 is largely possible now, although tooling is lacking. I think further discussion of 1 should happen on Secure Code WG.

**Comment by infinity0** [comment]
Continuing off from last time, more than a year ago:

> @infinity0
> 
> Nice to hear we almost got rid of nondeterminism! The last bunch of path dependence looks like it comes from [compiler-rt](https://github.com/rust-lang/compiler-rt) asm files that are a part of [compiler-builtins](https://github.com/rust-lang-nursery/compiler-builtins) - I think the build-script that compiles them all is [here](https://github.com/rust-lang-nursery/compiler-builtins/blob/f5532b22b5d741f3ea207b5b07e3e1ca63476f9b/build.rs#L4435).
> 
> I think it's just that we don't pass debuginfo remapping all the way down to gcc.

The cc crate now honours CFLAGS and CPPFLAGS so this problem has gone away and I can confirm the libcompiler_builtins builds reproducibly independently of build-path including the C code.

However there is a new regression, dylib_metadata is no longer reproducible. (rlib_metadata is fine.) Please see these samples:

- [librustc_codegen_llvm-llvm.so.rustc.1](https://people.debian.org/~infinity0/rust/librustc_codegen_llvm-llvm.so.rustc.1)
- [librustc_codegen_llvm-llvm.so.rustc.2](https://people.debian.org/~infinity0/rust/librustc_codegen_llvm-llvm.so.rustc.2)

These files were extract from the ".rustc" section of the respective ".so" files. I don't know how to parse the binary data into a human-readable format, so I can't continue further by myself. Someone needs to explain it a bit more.

**Comment by infinity0** [comment]
I think this is the last piece so if someone can figure this out in the next month (before Mar 12) we could in theory get a reproducible `rustc` into the next Debian Stable, if that motivates anyone to work on it.

**Comment by tarcieri** [comment]
As a side note to anyone interested in creating reproducible builds of libraries/apps authored in Rust, I've created a `cargo-repro` project in the Secure Code WG as a place to focus efforts on creating a Cargo-driven workflow for reproducible builds:

https://github.com/rust-secure-code/cargo-repro

(Apologies if this is off-topic for this particular issue, but I thought there might be some interest overlap in the people subscribed)

**Comment by jgalenson** [comment]
I managed to get reproducible builds of rustc (varying mostly only the source directory).  I have proof of concept patches and a script that uses them to make two identical builds at https://github.com/jgalenson/reproducible-rustc.  The patches are not ready to upstream immediately, and I'd greatly appreciate advice or help doing so.  Some are hackier than others. :)

I have more details and the patches themselves in the repository, but I needed to make the following changes:

- To pass `--remap-path-prefix` to all rustc instances, I used the CARGO_REGISTRY idea proposed in https://github.com/rust-lang/cargo/issues/5505.  I initially tried adding the flag in `.cargo/config` and the RUSTFLAGS environment variable, but for some reason neither quite worked (a few files were compiled without it).

- I needed to modify rustc to avoid hashing the `--sysroot` argument, which changes as the source directory changes.  My current patch is a bit ugly, as it creates an otherwise unused remapped sysroot option and hashes that instead of the original.  This is not quite correct, as the hash will remain unchanged if both the sysroot and remapped path change.  I also have a patch that modifies the hash function so the caller can choose whether or not to hash the sysroot, which is also a bit ugly.  Advice on improving this is definitely appreciated.

- When librustc_llvm compiles parts of LLVM, it uses the `__FILE__` macro, which contains the absolute path.  I modified it to pass -DNDEBUG (if the compile options allow it) from the bootstrap process, which avoids using `__FILE__`.

- Building compiler-rt (as part of libcompiler_builtins) also uses `__FILE__`, but unfortunately it doesn't have an alternative version defined.  So as a hack, I removed the `__FILE__`.  A better solution would be to use DNDEBUG or something like it (as above) or remap it once https://reviews.llvm.org/D49466 lands.

Anyway, I hope this helps others build rustc reproducibly, and I'd be happy for feedback on getting these patches upstreamed.

**Comment by daym** [comment]
@jgalenson: I guess it can't hurt.

But why not just use a build container and have the file names inside be fixed? I mean I'm sure there are use cases left where you can't use a build container (so it's good if your patches make it in), but all major operating systems support them now, most of them even for unprivileged users. So you can simplify your life a lot by calling `unshare`.

We do that and not all rusts in the chain starting with mrustc 0.8.0 (i.e. mrustc 0.8.0 -> rust 1.19.0 -> rust 1.20.0 -> [...] -> rust 1.36.0) build reproducibly even then, so I suspect yours doesn't either.

**Comment by jgalenson** [comment]
Yes, my patches don't affect older code, so the bootstrapping problem remains.  However, my goal was to make it easier to get reproducible builds going forward, so that techniques like `unshare` aren't needed past some point and it's easier to reproduce builds and detect reproducibility regressions.

**Comment by jgalenson** [comment]
Now that #63505, #63558, and #63621 have merged, rustc builds are almost completely reproducible for me (at least across different directories).  And now that rust-lang-nursery/compiler-builtins/pull/309 has merged, once it is pulled into a compiler-builtins release that rustc uses, the rustc build for me will be fully reproducible when built with gcc 8.

Additionally, with #63352 merged, I managed to build reproducibly all the crates that were failing in the [Debian dashboard](https://tests.reproducible-builds.org/debian/unstable/amd64/index_FTBR.html).

**Comment by tmandry** [comment]
Now that some effort has gone into this, I would love to lock it in with a check in Rust's nightly builders that a stage 3 build is identical to the stage 2 build. It wouldn't catch every kind of regression, but should help quite a bit I would think.

That said, I'm not sure what the impact would be on resources.

**Comment by Mark-Simulacrum** [comment]
We already run a full-bootstrap build (and it's not even the longest builder, IIRC), so just hashing the artifacts and such shouldn't be too hard. I am willing to review a PR that attempts to do this, though I do not personally have the time to do so.

(What you actually want is to compare stage 1 and stage 2 builds, with full-bootstrap enabled -- stage 3 is a bit nonsensical in all configurations).

**Comment by bjorn3** [comment]
I think the ci should also move the src and target dir between building stage 1 and building stage 2 for the repro test.

**Comment by infinity0** [comment]
Does it really make sense to compare stage1 with stage2? stage1 will be built with stage0 which normally would be a *different version* of the rustc compiler i.e. the previous one. So they would be expected to behave differently and therefore compile the same thing differently. Also the description for `full-bootstrap` sounds like it is the same thing as stage3, so I'm not sure why you say "stage 3 is a bit nonsensical".

In the [double-diverse-compilation proof-of-concept](https://github.com/bmwiedemann/ddcpoc) we did for the reproducible builds project a couple years back, we would typically expect stage2 and stage3 to be the same, not stage1. Although pgcc was buggy and somehow would only reach a fixed-point with stage3/stage4.

**Comment by infinity0** [comment]
Sadly 1.39.0 did not reproduce for me when varying the build-path, could you take a look @jgalenson? I believe all the PRs you mentioned above should be part of that version and I also added the compiler-builtins patch specifically to Debian's rustc. The reproducible builds project website has a [summary diff](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/diffoscope-results/rustc.html) but I've uploaded a [more detailed version here](https://people.debian.org/~infinity0/rust/diff.html/). If you enable javascript you can click or shift-click the -/+ to expand/collapse (recursively with shift) the different entries.

**Comment by bjorn3** [comment]
The first string difference is at https://tests.reproducible-builds.org/debian/dbd/unstable/amd64/rustc_1.39.0+dfsg1-2.diffoscope.html#libstd-rust--.--_-.--.--dfsg---_amd--.deb---data.tar.xz---data.tar---.-usr-lib-x--_---linux-gnu-librustc_macros-ce-d-----e------.so---readelf---wide---decompress---hex-dump-.rodata---.

**Comment by infinity0** [comment]
I've also uploaded the build artifacts [here](https://people.debian.org/~infinity0/rust/1.res/) and [here](https://people.debian.org/~infinity0/rust/2.res/). If you're not on a Debian system you can extract `.deb` files directly with `ar xf`.

**Comment by infinity0** [comment]
Also we use a slightly customised [config.toml](https://salsa.debian.org/rust-team/rust/blob/debian/sid/debian/config.toml.in). In particular, `debuginfo-level` and `debuginfo-level-std` are both 2.

**Comment by jgalenson** [comment]
First of all, thanks for looking into this!  That dashboard should be a great way to keep tabs on this and spot regressions (once we get it greener).

I wouldn't actually expect builds with debuginfo to be deterministic across different build paths.  Can you disable that line and see what changes?  And if that does help, do you need it enabled?

In addition, from that config.toml file it looks like you're passing --remap-path-prefix via RUSTFLAGS (in the rules file).  I remember that didn't work well for me, and so I instead used the remap-debuginfo config.toml file (see https://github.com/jgalenson/reproducible-rustc/commit/cdf504e5f0933a141b9fdcb451d213f272104a2f for a brief explanation of this).  Can you try that as well?

**Comment by infinity0** [comment]
Hey thanks for the quick response.

> I wouldn't actually expect builds with debuginfo to be deterministic across different build paths. Can you disable that line and see what changes? And if that does help, do you need it enabled?

I will try it, give me a few days. However as I remember, one of the main reason for that flag even existing in the first place is for debuginfo - see #41555 and related issues - so it really ought to be working, and I seem to remember that it worked fine before.

> --remap-path-prefix via RUSTFLAGS (in the rules file). I remember that didn't work well for me, 

Yes, I fixed this recently in #66834 and included this patch in this Debian version. If I `rg` the build logs I can see this flag being passed fine.

**Comment by infinity0** [comment]
> The first string difference is at https://tests.reproducible-builds.org/debian/dbd/unstable/amd64/rustc_1.39.0+dfsg1-2.diffoscope.html#libstd-rust--.--_-.--.--dfsg---_amd--.deb---data.tar.xz---data.tar---.-usr-lib-x--_---linux-gnu-librustc_macros-ce-d-----e------.so---readelf---wide---decompress---hex-dump-.rodata---.

This comes from code in the syn crate:

~~~~
$ rg "unknown.delimiter"
vendor/syn-0.15.35/src/token.rs
942:            _ => panic!("unknown delimiter: {}", s),
~~~~

Then again [some other stuff](https://people.debian.org/~infinity0/rust/diff.html/#libstd-rust-dev_-.--.--dfsg---_amd--.deb---data.tar.xz---data.tar---.-usr-lib-rustlib-x--_---unknown-linux-gnu-lib-libfmt_macros-eac---b-a-------.rlib---rust.metadata.bin) seems unrelated to build paths, so maybe someone added a nondeterministic hash table again.

Not sure if many people know this but 1.19.0 was already "nearly reproducible" (see my comments further up from 2 years ago) and so it seems adding some integration tests is really vital to actually maintaining this property.

**Comment by jgalenson** [comment]
Ever since I first got my builds to reproduce, I've been running a script daily that builds rustc twice and compares the output, and it hasn't regressed yet.  Having this integrated into the official test suite is of course a much better solution, but at least with my setup there haven't been any regressions so far.

I'll look into getting this to work with debuginfo next week.

**Comment by tarcieri** [comment]
@jgalenson awesome! I'm working on a project for BFT consensus-driven proofs of reproducible builds, and that'd make an amazing test case: https://github.com/iqlusioninc/synchronicity

**Comment by jgalenson** [comment]
I was able to get reproducible builds of rustc with debuginfo by upgrading to a newer version of the compiler_builtins crate that contains https://github.com/rust-lang/compiler-builtins/commit/ca423fe49a9cb90baf5dd3cfefc13c11d6c1eb02 (so at least 0.1.20 should work, although I used 0.1.22).  With that, I no longer needed my last remaining [patch](https://github.com/jgalenson/reproducible-rustc/blob/master/patches/llvm-0001-Hackily-disable-the-__FILE__-macro-in-compiler-rt.patch).

I also had to use a newer C++ compiler that supports the `-ffile-prefix-map` argument.  gcc-8 does, but unfortunately using it did not work (it seems to embed source paths into files generated from assembly).  However, since https://reviews.llvm.org/D49466 was merged, Clang supports that option now, so using a random nightly build of LLVM from a few days ago worked.

That's a convoluted setup, but @infinity0, can you reproduce it?  Your diff seemed to contain some other differences mine did not, so some things might remain.

@tarcieri that seems pretty cool.  Have you gotten it working yet?

**Comment by infinity0** [comment]
@jgalenson I'm already including that compiler-builtins `-ffile-prefix-map` patch in Debian's rustc and it doesn't work, though we're using GCC not LLVM. What exactly is this issue with "it seems to embed source paths into files generated from assembly" you are talking about? I don't remember seeing that last time I was working on GCC reproducibility a few years ago.

[tests.r-b.org results for rustc 1.39](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/diffoscope-results/rustc.html) are out the results are pretty similar to mine above - unreproducible including stuff relating to the `syn` crate.

**Comment by infinity0** [comment]
I should mention that the differences include more that just C++ compiler output, which is what you seem to be suggesting is the last remaining thing, from your previous comment. My own testing and tests.r-b.org's testing indicates otherwise. At least the syn crate is embedding build paths.

**Comment by jgalenson** [comment]
I investigated the gcc bug I mentioned.  If I take a trivial .c file and compile it with `-g` and  `-ffile-prefix-map`, I don't see source paths in the output.  But if I use the same command line for a trivial assembly file, I do see source paths in the output.  Those go away if I change `-ffile-prefix-map` to `-fdebug-prefix-map`.  I see this with gcc 8.3 but not with recent LLVM builds.  Can you reproduce that?

This comes up because compiler-builtins [includes](https://github.com/rust-lang/compiler-builtins/blob/master/build.rs#L249) a few .S files.  This miscompilation then causes a number of other crates to be non-reproducible, including rustc_{driver,llvm} and cargo.  Those all go away one I use Clang instead of gcc.  I don't, however, see a difference in syn.

The tests.r-b.org link you posted no longer works for me because it seems to fail to build.  But if you can send me the config.toml you're using, I'll see if I can reproduce it myself.  In addition, you can try using my [config](https://github.com/jgalenson/reproducible-rustc/blob/master/repro.sh#L34) and adding in `debuginfo-level{,-std,-tools} = 2` and using a recent Clang.

**Comment by comex** [comment]
Filed https://gcc.gnu.org/bugzilla/show_bug.cgi?id=93371 for the GCC bug.

**Comment by comex** [comment]
Filed https://gcc.gnu.org/bugzilla/show_bug.cgi?id=93371 for the GCC bug.  As a short-term workaround, compiler-builtins could pass both `-ffile-prefix-map` and `-fdebug-prefix-map`...

**Comment by infinity0** [comment]
I didn't get around to looking into the above issues in more detail, but I just noticed that [tests.r-b.org for rustc](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/rustc.html) has been showing "reproducible" for 1.44.1 including under varying build paths (unstable, experimental)!

1.45.0 is in-progress and we should have results in just a few more days.

**Comment by sanxiyn** [comment]
We should use individual issues tagged A-reproducibility going forward. Closing.

**Comment by infinity0** [comment]
OK, well results for 1.45.0 say "unreproducible" again, and the auto-generated diff did not complete after 2 hours. Shall I open an issue for CI integration tests on this? If I understand correctly there are concerns about it taking up too much resources, or something.

**Comment by sanxiyn** [comment]
Yes, I think we should open a new issue for reproducible build CI. I also think it would be impractical for a while due to computational demand, but then I am not a member of infra team.
